src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs

```diff
@@ -46,18 +48,63 @@ public class PersistentReceiptStorage : IReceiptStorage
             _receiptsRecovery = receiptsRecovery ?? throw new ArgumentNullException(nameof(receiptsRecovery));
             _blocksDb = _database.GetColumnDb(ReceiptsColumns.Blocks);
             _transactionDb = _database.GetColumnDb(ReceiptsColumns.Transactions);
-            _blockFinder = blockFinder ?? throw new ArgumentNullException(nameof(blockFinder));
+            _blockTree = blockTree ?? throw new ArgumentNullException(nameof(blockTree));
             _storageDecoder = storageDecoder ?? ReceiptArrayStorageDecoder.Instance;
+            _receiptConfig = receiptConfig ?? throw new ArgumentNullException(nameof(receiptConfig));
 
             byte[] lowestBytes = _database.Get(Keccak.Zero);
             _lowestInsertedReceiptBlock = lowestBytes is null ? (long?)null : new RlpStream(lowestBytes).DecodeLong();
             _migratedBlockNumber = Get(MigrationBlockNumberKey, long.MaxValue);
+
+            _blockTree.BlockAddedToMain += BlockTreeOnBlockAddedToMain;
+        }
+
+        private void BlockTreeOnBlockAddedToMain(object? sender, BlockReplacementEventArgs e)
+        {
+            // Dont block main loop
+            Task.Run(() =>
+            {
+                Block newMain = e.Block;
+
+                // Move finalized block's tx index to use block number
+                if (_receiptConfig.CompactTxIndex)
+                {
+                    Block newFinalized = _blockTree.FindBlock(newMain.Number - Reorganization.MaxDepth);
+                    if (newFinalized != null)
+                    {
+                        EnsureCanonical(newFinalized);
+                    }
+                }
+
+                // Delete old tx index
+                if (_receiptConfig.TxLookupLimit > 0)
+                {
+                    Block newOldTx = _blockTree.FindBlock(newMain.Number - _receiptConfig.TxLookupLimit.Value);
+                    if (newOldTx != null)
+                    {
+                        ClearTxIndexForBlock(newOldTx);
+                    }
+                }
```

Comment:
Wouldn't it be simpler to just use a block number all the time? And removed reorganized transactions.

Reply:
Fair point